File: baekjoon/cpp/2110.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int main() {
	cin.tie(0);
	ios::sync_with_stdio(0);

	int n, c;
	cin >> n >> c;

	vector<int> v;
	for (int i = 0;i < n;i++) {
		int x;
		cin >> x;
		v.push_back(x);
	}

	sort(v.begin(), v.end());

	int l = 1; // 공유기 사이 거리 최소값
	int r = v[n - 1] - v[0]; // 공유기 사이 거리 최대값
	int ans = 0; // 인접 공유기 사이 최대 거리

	while (l <= r) {
		int m = (l + r) / 2; // 인접 공유기 사이 최대 거리 후보값

		// 첫 번째 집에는 반드시 공유기를 설치한다.
		int cnt = 1;
		int s = v[0]; // 직전 공유기 설치 위치
		for (int i = 1;i < n;i++) {
			if (v[i] - s >= m) {
				cnt++;
				s = v[i];
			}
		}

		if (cnt >= c) {
			ans = m;
			l = m + 1;
		}
		else {
			r = m - 1;
		}
	}

	cout << ans;
}

File: baekjoon/python/2110.py
import sys

n, c = map(int,sys.stdin.readline().split())

arr = [int(sys.stdin.readline()) for _ in range(n)]
arr.sort()

l=0 # 공유기 사이 거리 최소값
r=arr[-1]-arr[0] # 공유기 사이 거리 최대값
ans=0 # 인접한 공유기 사이 최대 거리

while l<=r:
    m=(l+r)//2 # 공유기 간 최대 거리 후보값
    
    # 첫 번째 집에는 공유기를 반드시 설치한다.
    prev=arr[0] # 직전에 공유기를 설치한 잡
    cnt=1 # 설치한 공유기의 수
    for i in range(1, n):
        if arr[i]-prev>=m:
            cnt+=1
            prev=arr[i]

    if cnt>=c:
        ans=m
        l=m+1
    else:
        r=m-1

print(ans)